packages/repository/src/relations/relation.helpers.ts

```diff
@@ -48,3 +58,77 @@ export async function findByForeignKeys<
 }
 
 type StringKeyOf<T> = Extract<keyof T, string>;
+
+/**
+ * Returns model instances that include related models that have a registered
+ * resolver.
+ *
+ * @param targetRepository - The target repository where the model instances are found
+ * @param entities - An array of entity instances or data
+ * @param filter - Query filter
+ * @param options - Options for the operations
+ */
+
+export async function includeRelatedModels<
+  T extends Entity,
+  Relations extends object = {}
+>(
+  targetRepository: EntityCrudRepository<T, unknown, Relations>,
+  entities: T[],
+  filter?: Filter<T>,
+  options?: Options,
+): Promise<(T & Relations)[]> {
+  const result = entities as (T & Relations)[];
+  const include = filter && filter.include;
+  if (!include) return result;
+
+  const invalidInclusions = include.filter(
+    i => !isInclusionAllowed(targetRepository, i),
+  );
+  if (invalidInclusions.length) {
+    const msg =
+      'Invalid "filter.include" entries: ' +
+      invalidInclusions.map(i => JSON.stringify(i)).join('; ');
+    const err = new Error(msg);
+    Object.assign(err, {
+      code: 'INVALID_INCLUSION_FILTER',
+    });
+    throw err;
+  }
+
+  const resolveTasks = include.map(async i => {
+    const relationName = i.relation;
+    const resolver = targetRepository.inclusionResolvers.get(relationName)!;
+    const targets = await resolver(entities, i, options);
+
+    for (const ix in result) {
+      const src = result[ix];
+      (src as AnyObject)[relationName] = targets[ix];
+    }
+  });
+
+  await Promise.all(resolveTasks);
```

Comment:
I am not sure is it good to execute them in parallel...since the tasks operates on the same object.

Reply:
> I am not sure is it good to execute them in parallel...since the tasks operates on the same object.

We are just reading it, not modifying it. So I guess it should be fine?

Reply:
Each task is resolving a single relation. Even if there were multiple relations configured to use the same navigational property (which is a programmer error, IMO), then this code would be still safe to execute in parallel. JavaScript is single threaded, parallel execution means that we fire all database queries at the same time, the callback (code executed after `await resolver(...)`) is always run one at a time. If we switched to serial execution, then we will be running one DB query at a time, slowing down the process.

Reply:
Thank you @agnes512 @bajtos 👍 fair enough.